### ISSUE TITLE
Cmdlets may ignore parameter aliases in case of naming collisions with other parameters; e.g., "Get-ChildItem -d"

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#4739]
A parameter alias is designed to give an _unambiguous_ alternative name - typically shorter than the original name - to a parameter.

For instance, `Get-ChildItem`'s `-Directory` parameter (switch) has two aliases: `-ad` and `-d`.

Thus, both `Get-ChildItem -ad` and `Get-ChildItem -d` should work the same as `Get-ChildItem -Directory`.

However, `Get-ChildItem -d` is currently misinterpreted as the `-Depth` parameter.

The same problem _may_ affect other cmdlets analogously.

Steps to reproduce
------------------

```powershell
Get-ChildItem -ad
Get-ChildItem -d
```

Expected behavior
-----------------

Both commands should list the subdirectories of the current directory.

Actual behavior
---------------

`Get-ChildItem -ad` succeeds, but `Get-ChildItem -d` fails with the following error:

```none
Get-ChildItem : Missing an argument for parameter 'Depth'. Specify a parameter of type 'System.UInt32' and try again.
```

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta.6 on macOS 10.12.6
PowerShell Core v6.0.0-beta.6 on Ubuntu 16.04.3 LTS
PowerShell Core v6.0.0-beta.6 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
Windows PowerShell v5.1.15063.502 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
```

**Comment by PetSerAl** [comment]
`-Directory` is not a parameter of `Get-ChildItem` cmdlet, while `-Depth` is. `-Directory` is dynamic parameter of `FileSystem` provider. Thus when PowerShell bind static parameters of `Get-ChildItem` cmdlet `-d` is unambiguously resolved to `-Depth`. PowerShell does not try to find better match for already bound static parameters.

```powershell
function f {
    [CmdletBinding()]
    param($A1)
    dynamicparam {
        Write-Host "A1 paramter have value '$A1'"
        $DynParams = [System.Management.Automation.RuntimeDefinedParameterDictionary]::new()
        $A2Param = [System.Management.Automation.RuntimeDefinedParameter]::new('A2', [object], ([Parameter]::new(), [Alias]::new('A', 'B')))
        $DynParams.Add('A2', $A2Param)
        $DynParams
    }
    end {
        "A1 parameter: '$A1'"
        "A2 parameter: '$($A2Param.Value)'"
    }
}
f -A 42
f -B 42
```

**Comment by mklement0** [comment]
As usual, thanks for digging deeper, @PetSerAl.

Given the current behavior, the conclusions are:

* Remove the `d` alias from the `Directory` parameter declaration, because it is effectively _inaccessible_.

* Update the documentation accordingly.

Leaving it at that is an option, but the bigger question is:

Can dynamic parameters be considered alongside the static ones too when looking for ambiguity / unambiguous aliases?

From a UX perspective, this seems preferable.

Or are there good (a) conceptual or (b) technical reasons not to do that?

**Comment by mklement0** [comment]
Closed in favor of #14700.
